File: apps/android/app/src/main/kotlin/com/dropbeam/FileTransfer.kt
package com.dropbeam

import android.content.Context
import android.net.wifi.p2p.WifiP2pDevice
import android.net.wifi.p2p.WifiP2pInfo
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import org.json.JSONArray
import org.json.JSONObject
import org.webrtc.IceCandidate
import org.webrtc.PeerConnection
import org.webrtc.SessionDescription
import java.io.File
import java.util.UUID

// High-level transfer coordinator.
// Route: WiFi Direct (same LAN) → WebRTC (remote/fallback)

enum class TransportKind { NEARBY, WEBRTC }

sealed class TransferEvent {
    data class Discovered(val device: WifiP2pDevice) : TransferEvent()
    data class Progress(val name: String, val bytes: Long, val total: Long) : TransferEvent()
    data class Complete(val name: String, val file: File?) : TransferEvent()
    data class Error(val error: Throwable) : TransferEvent()
    data class RoomCode(val code: String) : TransferEvent()
}

class FileTransfer(
    private val context: Context,
    private val signalingUrl: String,
    private val displayName: String,
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO + SupervisorJob())
) {
    private val _events = MutableSharedFlow<TransferEvent>(extraBufferCapacity = 64)
    val events: SharedFlow<TransferEvent> = _events.asSharedFlow()

    private val nearby = NearbyTransfer(context, scope)
    private val webrtc by lazy { WebRTCFallback(context, scope) }
    private var signaling: SignalingClient? = null

    private var pendingFiles: List<File> = emptyList()
    private var remotePeerId: String? = null
    private var peerId = UUID.randomUUID().toString()

    init {
        scope.launch { nearby.events.collect { handleNearbyEvent(it) } }
    }

    // MARK: - Public API

    fun startDiscovery() = nearby.start()
    fun stopDiscovery() = nearby.stop()

    fun connectNearby(device: WifiP2pDevice) = nearby.connect(device)

    fun sendNearby(files: List<File>, info: WifiP2pInfo) {
        pendingFiles = files
        if (info.isGroupOwner) {
            // We are the GO — start a server socket and wait for peer
            // (peer will connect to us; we send after their connection)
        } else {
            nearby.sendFiles(files, info.groupOwnerAddress.hostAddress ?: return)
        }
    }

    fun sendRemote(files: List<File>, roomCode: String) {
        pendingFiles = files
        connectSignaling { client ->
            client.joinRoom(code = roomCode, device = myDevice())
        }
    }

    fun receiveRemote() {
        connectSignaling { client ->
            client.createRoom(myDevice())
        }
    }

    fun close() {
        stopDiscovery()
        webrtc.close()
        signaling?.disconnect()
        scope.cancel()
    }

    // MARK: - Private

    private fun myDevice() = DeviceInfo(
        name = displayName, kind = "android", peerId = peerId
    )

    private fun connectSignaling(setup: (SignalingClient) -> Unit) {
        val client = SignalingClient(signalingUrl, scope)
        signaling = client

        scope.launch {
            client.events.collect { event ->
                when (event) {
                    is ServerEvent.RoomCreated -> {
                        _events.emit(TransferEvent.RoomCode(event.code))
                        setupWebRTCAsReceiver()
                    }
                    is ServerEvent.RoomJoined -> {
                        remotePeerId = event.peerId
                        setupWebRTCAsSender()
                    }
                    is ServerEvent.PeerJoined -> {
                        remotePeerId = event.peerId
                        setupWebRTCAsSender()
                    }
                    is ServerEvent.Signal -> handleSignal(event.from, event.data)
                    is ServerEvent.Error -> _events.emit(TransferEvent.Error(Exception(event.message)))
                    else -> {}
                }
            }
        }

        client.connect()
        setup(client)
    }

    private fun setupWebRTCAsSender() {
        scope.launch { webrtc.events.collect { handleWebRTCEvent(it) } }
        webrtc.setupAsSender(WebRTCFallback.defaultIceServers())
        webrtc.createOffer()
    }

    private fun setupWebRTCAsReceiver() {
        scope.launch { webrtc.events.collect { handleWebRTCEvent(it) } }
        webrtc.setupAsReceiver(WebRTCFallback.defaultIceServers())
    }

    private fun handleWebRTCEvent(event: WebRTCEvent) {
        when (event) {
            is WebRTCEvent.ICECandidate -> {
                val target = remotePeerId ?: return
                signaling?.signal(target, JSONObject().apply {
                    put("type", "ice-candidate")
                    put("candidate", event.candidate.sdp)
                    put("sdpMid", event.candidate.sdpMid)
                    put("sdpMLineIndex", event.candidate.sdpMLineIndex)
                })
            }
            is WebRTCEvent.LocalDescription -> {
                val target = remotePeerId ?: return
                signaling?.signal(target, JSONObject().apply {
                    put("type", if (event.sdp.type == SessionDescription.Type.OFFER) "offer" else "answer")
                    put("sdp", event.sdp.description)
                })
            }
            is WebRTCEvent.StateChanged -> {
                if (event.state == PeerConnection.PeerConnectionState.CONNECTED) {
                    scope.launch { startWebRTCSend() }
                }
            }
            is WebRTCEvent.DataReceived -> {
                if (event.isBinary) {
                    scope.launch { _events.emit(TransferEvent.Progress("incoming", event.data.size.toLong(), -1)) }
                } else {
                    processControlMessage(String(event.data))
                }
            }
            is WebRTCEvent.Error -> scope.launch {
                _events.emit(TransferEvent.Error(Exception(event.message)))
            }
        }
    }

    private fun handleSignal(from: String, data: JSONObject) {
        when (data.optString("type")) {
            "offer" -> {
                remotePeerId = from
                val sdp = SessionDescription(SessionDescription.Type.OFFER, data.getString("sdp"))
                webrtc.setRemoteDescription(sdp)
                webrtc.createAnswer()
            }
            "answer" -> {
                val sdp = SessionDescription(SessionDescription.Type.ANSWER, data.getString("sdp"))
                webrtc.setRemoteDescription(sdp)
            }
            "ice-candidate" -> {
                val candidate = IceCandidate(
                    data.getString("sdpMid"),
                    data.getInt("sdpMLineIndex"),
                    data.getString("candidate")
                )
                webrtc.addIceCandidate(candidate)
            }
        }
    }

    private suspend fun startWebRTCSend() {
        val files = pendingFiles
        if (files.isEmpty()) return
        pendingFiles = emptyList()

        val manifestArr = JSONArray()
        files.forEach { f ->
            manifestArr.put(JSONObject().apply {
                put("name", f.name); put("size", f.length()); put("id", UUID.randomUUID().toString())
            })
        }
        webrtc.sendControl(JSONObject().apply {
            put("type", "manifest"); put("files", manifestArr)
        }.toString())

        files.forEachIndexed { fileIdx, file ->
            streamFileWebRTC(file, fileIdx)
        }
        webrtc.sendControl("""{"type":"complete"}""")
    }

    private suspend fun streamFileWebRTC(file: File, fileId: Int) {
        val chunkSize = 256 * 1024
        file.inputStream().use { fis ->
            var chunkIndex = 0
            val buf = ByteArray(chunkSize)
            var n: Int
            while (fis.read(buf).also { n = it } != -1) {
                val payload = buf.copyOf(n)
                val frame = ByteArray(16 + n)
                frame[0] = 0xDB.toByte()
                frame[1] = 1
                frame[2] = 0
                frame[3] = 0
                writeInt32BE(frame, 4, fileId)
                writeInt32BE(frame, 8, chunkIndex)
                writeInt32BE(frame, 12, n)
                payload.copyInto(frame, 16)
                webrtc.sendData(frame, chunkIndex % webrtc.lanes)
                _events.emit(TransferEvent.Progress(file.name, (chunkIndex.toLong() + 1) * chunkSize, file.length()))
                chunkIndex++
            }
        }
    }

    private fun processControlMessage(text: String) {
        val json = runCatching { JSONObject(text) }.getOrNull() ?: return
        when (json.optString("type")) {
            "manifest" -> { /* host app can display file list */ }
            "complete" -> scope.launch { _events.emit(TransferEvent.Complete("transfer", null)) }
        }
    }

    private suspend fun handleNearbyEvent(event: NearbyEvent) {
        when (event) {
            is NearbyEvent.PeersDiscovered -> event.peers.forEach {
                _events.emit(TransferEvent.Discovered(it))
            }
            is NearbyEvent.ReceiveProgress -> _events.emit(
                TransferEvent.Progress(event.name, event.bytes, event.total)
            )
            is NearbyEvent.ReceiveComplete -> _events.emit(
                TransferEvent.Complete(event.name, event.file)
            )
            is NearbyEvent.TransferError -> _events.emit(TransferEvent.Error(event.error))
            else -> {}
        }
    }

    private fun writeInt32BE(buf: ByteArray, offset: Int, value: Int) {
        buf[offset] = (value shr 24).toByte()
        buf[offset + 1] = (value shr 16).toByte()
        buf[offset + 2] = (value shr 8).toByte()
        buf[offset + 3] = value.toByte()
    }
}

File: apps/ios/DropBeam/FileTransfer.swift
import Foundation
import MultipeerConnectivity

// High-level transfer coordinator.
// Route: MultipeerConnectivity (same LAN) → WebRTC (remote/fallback)

enum TransportKind { case nearby, webrtc }

struct TransferProgress {
    let fileName: String
    let bytesTransferred: Int64
    let totalBytes: Int64
    var fraction: Double { totalBytes > 0 ? Double(bytesTransferred) / Double(totalBytes) : 0 }
}

protocol FileTransferDelegate: AnyObject {
    func fileTransfer(_ ft: FileTransfer, didUpdateProgress progress: TransferProgress)
    func fileTransfer(_ ft: FileTransfer, didCompleteFile name: String, savedTo url: URL?)
    func fileTransfer(_ ft: FileTransfer, didFail error: Error)
    func fileTransfer(_ ft: FileTransfer, didDiscover peer: MCPeerID)
}

final class FileTransfer {
    weak var delegate: FileTransferDelegate?

    private let displayName: String
    private let signalingURL: URL
    private lazy var nearby = NearbyTransfer(displayName: displayName)
    private lazy var webrtc = WebRTCFallback()
    private var signaling: SignalingClient?
    private var activeTransport: TransportKind = .nearby

    init(displayName: String, signalingURL: URL) {
        self.displayName = displayName
        self.signalingURL = signalingURL
        nearby.delegate = self
    }

    // MARK: - Public API

    func startDiscovery() {
        nearby.start()
    }

    func stopDiscovery() {
        nearby.stop()
    }

    /// Send files to a nearby (same-LAN) peer.
    func sendNearby(urls: [URL], to peer: MCPeerID) {
        activeTransport = .nearby
        nearby.invite(peer)
        // sendFiles is called after connection established in delegate
        pendingURLs = urls
        pendingPeer = peer
    }

    /// Send files over WebRTC (remote peer identified by room code).
    func sendRemote(urls: [URL], roomCode: String) {
        activeTransport = .webrtc
        pendingURLs = urls
        connectSignaling(asInitiator: false, roomCode: roomCode)
    }

    /// Join as receiver (show room code to sender).
    func receiveRemote(completion: @escaping (String) -> Void) {
        activeTransport = .webrtc
        connectSignaling(asInitiator: true, roomCode: nil, onCode: completion)
    }

    // MARK: - Private

    private var pendingURLs: [URL] = []
    private var pendingPeer: MCPeerID?

    private func connectSignaling(asInitiator: Bool, roomCode: String?,
                                  onCode: ((String) -> Void)? = nil) {
        let device = DeviceInfo(name: displayName, kind: "ios",
                                peerId: UUID().uuidString, version: 1)
        let client = SignalingClient(url: signalingURL)
        signaling = client

        client.onRoomCreated = { [weak self] _, code in
            onCode?(code)
            self?.setupWebRTCAsSender()
        }
        client.onRoomJoined = { [weak self] _, _ in
            self?.setupWebRTCAsReceiver()
        }
        client.on { [weak self] from, data in
            self?.handleSignal(from: from, data: data)
        }
        client.onError = { [weak self] msg in
            self?.delegate?.fileTransfer(self!, didFail: NSError(domain: "DropBeam", code: -1,
                                                                  userInfo: [NSLocalizedDescriptionKey: msg]))
        }

        client.connect()
        if asInitiator {
            client.createRoom(device)
        } else if let code = roomCode {
            client.joinRoom(code: code, device: device)
        }
    }

    private func setupWebRTCAsSender() {
        webrtc.delegate = self
        webrtc.setupAsSender(iceServers: WebRTCFallback.defaultIceServers())
        webrtc.onICECandidate = { [weak self] candidate in
            self?.signaling?.signal(to: "broadcast", data: [
                "type": AnyCodable("ice-candidate"),
                "candidate": AnyCodable(candidate.sdp),
                "sdpMid": AnyCodable(candidate.sdpMid ?? ""),
                "sdpMLineIndex": AnyCodable(Int(candidate.sdpMLineIndex))
            ])
        }
        webrtc.createOffer { [weak self] sdp, error in
            guard let self = self, let sdp = sdp else { return }
            self.signaling?.signal(to: "broadcast", data: [
                "type": AnyCodable("offer"),
                "sdp": AnyCodable(sdp.sdp)
            ])
        }
    }

    private func setupWebRTCAsReceiver() {
        webrtc.delegate = self
        webrtc.setupAsReceiver(iceServers: WebRTCFallback.defaultIceServers())
        webrtc.onICECandidate = { [weak self] candidate in
            self?.signaling?.signal(to: "broadcast", data: [
                "type": AnyCodable("ice-candidate"),
                "candidate": AnyCodable(candidate.sdp),
                "sdpMid": AnyCodable(candidate.sdpMid ?? ""),
                "sdpMLineIndex": AnyCodable(Int(candidate.sdpMLineIndex))
            ])
        }
    }

    private func handleSignal(from: String, data: [String: AnyCodable]) {
        guard let type = data["type"]?.value as? String else { return }
        switch type {
        case "offer":
            guard let sdpStr = data["sdp"]?.value as? String else { return }
            let sdp = RTCSessionDescription(type: .offer, sdp: sdpStr)
            webrtc.setRemoteDescription(sdp) { [weak self] _ in
                self?.webrtc.createAnswer { answer, _ in
                    guard let self = self, let answer = answer else { return }
                    self.signaling?.signal(to: from, data: [
                        "type": AnyCodable("answer"),
                        "sdp": AnyCodable(answer.sdp)
                    ])
                }
            }
        case "answer":
            guard let sdpStr = data["sdp"]?.value as? String else { return }
            let sdp = RTCSessionDescription(type: .answer, sdp: sdpStr)
            webrtc.setRemoteDescription(sdp) { _ in }
        case "ice-candidate":
            guard let candidateStr = data["candidate"]?.value as? String,
                  let mid = data["sdpMid"]?.value as? String,
                  let lineIndex = data["sdpMLineIndex"]?.value as? Int else { return }
            let candidate = RTCIceCandidate(sdp: candidateStr, sdpMLineIndex: Int32(lineIndex), sdpMid: mid)
            webrtc.addIceCandidate(candidate)
        default:
            break
        }
    }
}

// MARK: - NearbyTransferDelegate

extension FileTransfer: NearbyTransferDelegate {
    func nearbyTransfer(_ transfer: NearbyTransfer, didDiscover peer: MCPeerID,
                        info: [String: String]) {
        delegate?.fileTransfer(self, didDiscover: peer)
    }

    func nearbyTransfer(_ transfer: NearbyTransfer, didConnect peer: MCPeerID) {
        if let urls = Optional(pendingURLs), !urls.isEmpty, peer == pendingPeer {
            nearby.sendFiles(urls: urls, to: peer)
            pendingURLs = []
        }
    }

    func nearbyTransfer(_ transfer: NearbyTransfer, didDisconnect peer: MCPeerID) {}

    func nearbyTransfer(_ transfer: NearbyTransfer, didReceiveFile name: String,
                        progress: Double, url: URL?) {
        let p = TransferProgress(fileName: name, bytesTransferred: Int64(progress * 1e9),
                                 totalBytes: Int64(1e9))
        delegate?.fileTransfer(self, didUpdateProgress: p)
        if progress >= 1.0, let url = url {
            delegate?.fileTransfer(self, didCompleteFile: name, savedTo: url)
        }
    }

    func nearbyTransfer(_ transfer: NearbyTransfer, transferFailed error: Error) {
        delegate?.fileTransfer(self, didFail: error)
    }
}

// MARK: - WebRTCFallbackDelegate

extension FileTransfer: WebRTCFallbackDelegate {
    func webRTC(_ rtc: WebRTCFallback, didChangeState state: RTCPeerConnectionState) {
        if state == .connected && !pendingURLs.isEmpty {
            // Sender: begin streaming manifest + chunks via control/data channels
            // Full frame encoding mirrors TransferSender in packages/transfer
            startWebRTCSend()
        }
    }

    func webRTC(_ rtc: WebRTCFallback, didReceiveData data: Data) {
        // Decode frame header (16 bytes) + payload — same format as TS receiver
        processDataFrame(data)
    }

    func webRTC(_ rtc: WebRTCFallback, didReceiveControl text: String) {
        processControlMessage(text)
    }

    private func startWebRTCSend() {
        let urls = pendingURLs
        pendingURLs = []
        let manifest: [[String: Any]] = urls.map { url in
            let size = (try? FileManager.default.attributesOfItem(atPath: url.path)[.size] as? Int64) ?? 0
            return ["name": url.lastPathComponent, "size": size, "id": UUID().uuidString]
        }
        if let data = try? JSONSerialization.data(withJSONObject: ["type": "manifest", "files": manifest]),
           let text = String(data: data, encoding: .utf8) {
            webrtc.sendControl(text)
        }
        for (fileIdx, url) in urls.enumerated() {
            streamFile(url: url, fileId: UInt32(fileIdx))
        }
        webrtc.sendControl("{\"type\":\"complete\"}")
    }

    private func streamFile(url: URL, fileId: UInt32) {
        guard let fh = FileHandle(forReadingAtPath: url.path) else { return }
        defer { fh.closeFile() }
        var chunkIndex: UInt32 = 0
        let chunkSize = 256 * 1024
        while true {
            let data = fh.readData(ofLength: chunkSize)
            if data.isEmpty { break }
            var header = Data(count: 16)
            header[0] = 0xDB  // magic
            header[1] = 1     // version
            header[2] = 0     // flags
            header[3] = 0     // reserved
            header.withUnsafeMutableBytes { ptr in
                ptr.storeBytes(of: fileId.bigEndian, toByteOffset: 4, as: UInt32.self)
                ptr.storeBytes(of: chunkIndex.bigEndian, toByteOffset: 8, as: UInt32.self)
                ptr.storeBytes(of: UInt32(data.count).bigEndian, toByteOffset: 12, as: UInt32.self)
            }
            let frame = header + data
            webrtc.sendData(frame, lane: Int(chunkIndex) % webrtc.lanes)
            chunkIndex += 1
        }
    }

    private func processDataFrame(_ data: Data) {
        guard data.count >= 16 else { return }
        // Minimal receive — host app handles reassembly
        delegate?.fileTransfer(self, didUpdateProgress:
            TransferProgress(fileName: "incoming", bytesTransferred: Int64(data.count), totalBytes: -1))
    }

    private func processControlMessage(_ text: String) {
        guard let json = try? JSONSerialization.jsonObject(with: text.data(using: .utf8)!) as? [String: Any],
              let type = json["type"] as? String else { return }
        if type == "complete" {
            delegate?.fileTransfer(self, didCompleteFile: "transfer", savedTo: nil)
        }
    }
}

// WebRTC import placeholder — resolved by CocoaPods/SPM GoogleWebRTC
import class WebRTC.RTCSessionDescription
import class WebRTC.RTCIceCandidate
